lib/pbench/server/cache_manager.py

```diff
@@ -1433,3 +1505,115 @@ def delete(self, dataset_id: str):
+
+        @dataclass
+        class Candidate:
+            last_ref: float  # last reference timestamp
+            cache: Path  # Unpacked artifact directory
+
+        def reached_goal():
+            usage = shutil.disk_usage(self.cache_root)
+            available = float(usage.free) / float(usage.total) * 100.0
+            return bool(available >= goal_pct and usage.free >= goal_bytes)
+
+        if reached_goal():
+            return True
+
+        total_count = 0
+        reclaimed = 0
+        reclaim_failed = 0
+        self.logger.info(
+            "RECLAIM: looking for {}% or {} free",
+            goal_pct,
+            humanize.naturalsize(goal_bytes),
+        )
+
+        # Identify cached datasets by examining the cache directory tree
+        candidates = deque()
+        for d in self.cache_root.iterdir():
+            if not d.is_dir():
+                self.logger.warning("unexpected file in cache root: {}", d)
+                continue
+            total_count += 1
+            last_ref = 0.0
+            for f in d.iterdir():
+                if f.name == "last_ref":
+                    last_ref = f.stat().st_mtime
+                elif f.is_dir():
+                    candidates.append(Candidate(last_ref, f))
+
+        # Sort the candidates by last_ref timestamp, putting the oldest at
+        # the head of the queue. We'll flush each cache tree until we reach
+        # our goals.
+        candidates = sorted(candidates, key=lambda c: c.last_ref)
+        has_cache = len(candidates)
+        for candidate in candidates:
+            name = candidate.cache.name
+            cache_d = candidate.cache.parent
+            resource_id = cache_d.name
+
+            # Only if the dataset we're flushing has already been discovered,
+            # we want to update the Tarball object so that we don't break it.
+            # If it hasn't been discovered under this cache instance, that's
+            # fine because discovery will notice that the cache is empty.
+            if resource_id in self.datasets:
+                target = self.datasets[resource_id]
+            else:
+                target = None
+            try:
+                self.logger.info("RECLAIM: removing cache for {}", name)
+                with LockManager(cache_d / "lock", exclusive=True, wait=False):
+                    if target:
+                        target.cache_delete()
+                    else:
+                        shutil.rmtree(candidate.cache)
+                    reclaimed += 1
+                    if reached_goal():
+                        break
+            except OSError as e:
+                if e.errno in (errno.EAGAIN, errno.EACCES):
+                    self.logger.info(
+                        "RECLAIM {}: skipping because cache is locked", name
+                    )
+                    # Don't reclaim a cache that's in use
+                    continue
+                error = e
+                reclaim_failed += 1
```

Comment:
This `except` clause is intended to handle errors from `LockManager()`, right?...but, can `cache_delete()` or `rmtree()` raise either of these errors (particularly `EACCES`)?  If so, then the log message is misleading.

Reply:
Distinguishing the two cases makes the code messier. Since you keep pushing, I'm doing it anyway; although I'm still not convinced it's worth the trouble. 😦